app/questionnaire/placeholder_parser.py

```diff
@@ -134,3 +156,81 @@ def _resolve_value_source_list(
             else:
                 values.append(value)
         return values
+
+    def _get_routing_path_block_ids(self) -> list[str]:
+        return self._get_block_ids_for_calculated_summary_dependencies()
+
+    def _get_block_ids_for_calculated_summary_dependencies(self) -> list:
```

Comment:
This is almost identical to the one in questionnaire form. I suggest having one, returning something generic that both can make use of. If you give a list of section to ignore, you can make use of cache and then set the cache based on the return value. It should return a dict keyed by section id and then blocks (same as the cache itself).

In placeholder, it can then set the cache, in questionnaire form, it just needs to flatten the block ids.

Something like (not tested FYI):
```py
    def _get_block_ids_for_calculated_summary_dependencies(
        self, sections_to_ignore: Sequence | None = None
    ) -> dict[str, list[str]]:
        # Type ignore: Added to this method as the block will exist at this point
        blocks_id_by_section = {}

        sections_to_ignore = sections_to_ignore or []
        dependent_sections = self._schema.calculated_summary_section_dependencies_by_block.get(  # type: ignore
            self._location.section_id
        )
        if block_id := self._location.block_id:
            dependents = dependent_sections[block_id]

        else:
            dependents = {
                section
                for dependents in dependent_sections.values()
                if dependents
                for section in dependents
            }

        for section in dependents:
            if section in sections_to_ignore:
                continue

            key = (section, self._location.list_item_id)
            if key in self._path_finder.progress_store.started_section_keys():
                path = self._path_finder.routing_path(*key)
                blocks_id_by_section[section] = path.block_ids

        return blocks_id_by_section
```

then in placeholder, you can update the cache. The blocks ids you pass can just be a flattened list of the cached routing path ids. You can have a common method to do this, as you need to flatten in both form and parser.

Reply:
I've refactored these methods in-line with the above 👍